File: frontend/lib/instruments.ts
import type { Group } from "./api";

export interface BuiltinInstrument {
  slug: string;
  symbol: string; // yfinance ticker
  nameJa: string;
  group: Group;
}

/** Mirrors the backend registry — used by the strategy builder's quick-pick. */
export const BUILTIN_INSTRUMENTS: BuiltinInstrument[] = [
  { slug: "sp500", symbol: "^GSPC", nameJa: "S&P 500", group: "index" },
  { slug: "nasdaq", symbol: "^IXIC", nameJa: "NASDAQ総合", group: "index" },
  { slug: "dow", symbol: "^DJI", nameJa: "ダウ平均", group: "index" },
  { slug: "russell2000", symbol: "^RUT", nameJa: "ラッセル2000", group: "index" },
  { slug: "nikkei", symbol: "^N225", nameJa: "日経225", group: "index" },
  { slug: "ftse", symbol: "^FTSE", nameJa: "FTSE100", group: "index" },
  { slug: "dax", symbol: "^GDAXI", nameJa: "DAX", group: "index" },
  { slug: "hangseng", symbol: "^HSI", nameJa: "ハンセン", group: "index" },
  { slug: "vix", symbol: "^VIX", nameJa: "VIX恐怖指数", group: "index" },
  { slug: "XLK", symbol: "XLK", nameJa: "情報技術", group: "sector" },
  { slug: "XLC", symbol: "XLC", nameJa: "通信サービス", group: "sector" },
  { slug: "XLY", symbol: "XLY", nameJa: "一般消費財", group: "sector" },
  { slug: "XLP", symbol: "XLP", nameJa: "生活必需品", group: "sector" },
  { slug: "XLV", symbol: "XLV", nameJa: "ヘルスケア", group: "sector" },
  { slug: "XLF", symbol: "XLF", nameJa: "金融", group: "sector" },
  { slug: "XLI", symbol: "XLI", nameJa: "資本財", group: "sector" },
  { slug: "XLE", symbol: "XLE", nameJa: "エネルギー", group: "sector" },
  { slug: "XLB", symbol: "XLB", nameJa: "素材", group: "sector" },
  { slug: "XLU", symbol: "XLU", nameJa: "公益", group: "sector" },
  { slug: "XLRE", symbol: "XLRE", nameJa: "不動産", group: "sector" },
];

export const BUILTIN_SYMBOLS = new Set(BUILTIN_INSTRUMENTS.map((i) => i.symbol));

File: backend/app/instruments.py
"""Tradable instruments we expose: the 11 sector SPDR ETFs and major indices.

Each instrument has a URL-safe ``slug`` (indices use names like ``sp500``
because their yfinance ``symbol`` contains ``^``), a ``group`` so the frontend
can show sectors and indices in separate sections, and a ``currency`` used to
convert prices to JPY for the backtester.
"""

from dataclasses import dataclass


@dataclass(frozen=True)
class Instrument:
    slug: str
    symbol: str  # yfinance ticker
    name: str
    name_ja: str
    color: str
    group: str  # "sector" | "index"
    currency: str  # native quote currency, e.g. "USD"


SECTORS: list[Instrument] = [
    Instrument("XLK", "XLK", "Information Technology", "情報技術", "#3b82f6", "sector", "USD"),
    Instrument("XLC", "XLC", "Communication Services", "通信サービス", "#8b5cf6", "sector", "USD"),
    Instrument("XLY", "XLY", "Consumer Discretionary", "一般消費財", "#ec4899", "sector", "USD"),
    Instrument("XLP", "XLP", "Consumer Staples", "生活必需品", "#14b8a6", "sector", "USD"),
    Instrument("XLV", "XLV", "Health Care", "ヘルスケア", "#06b6d4", "sector", "USD"),
    Instrument("XLF", "XLF", "Financials", "金融", "#22c55e", "sector", "USD"),
    Instrument("XLI", "XLI", "Industrials", "資本財", "#f59e0b", "sector", "USD"),
    Instrument("XLE", "XLE", "Energy", "エネルギー", "#ef4444", "sector", "USD"),
    Instrument("XLB", "XLB", "Materials", "素材", "#a855f7", "sector", "USD"),
    Instrument("XLU", "XLU", "Utilities", "公益", "#64748b", "sector", "USD"),
    Instrument("XLRE", "XLRE", "Real Estate", "不動産", "#eab308", "sector", "USD"),
]

INDICES: list[Instrument] = [
    Instrument("sp500", "^GSPC", "S&P 500", "S&P 500", "#22c55e", "index", "USD"),
    Instrument("nasdaq", "^IXIC", "Nasdaq Composite", "NASDAQ総合", "#3b82f6", "index", "USD"),
    Instrument("dow", "^DJI", "Dow Jones", "ダウ平均", "#06b6d4", "index", "USD"),
    Instrument("russell2000", "^RUT", "Russell 2000", "ラッセル2000", "#a855f7", "index", "USD"),
    Instrument("nikkei", "^N225", "Nikkei 225", "日経225", "#ef4444", "index", "JPY"),
    Instrument("ftse", "^FTSE", "FTSE 100", "FTSE100", "#f59e0b", "index", "GBP"),
    Instrument("dax", "^GDAXI", "DAX", "DAX", "#eab308", "index", "EUR"),
    Instrument("hangseng", "^HSI", "Hang Seng", "ハンセン", "#ec4899", "index", "HKD"),
    Instrument("vix", "^VIX", "Volatility Index", "VIX恐怖指数", "#f97316", "index", "USD"),
]

ALL: list[Instrument] = SECTORS + INDICES
BY_SLUG: dict[str, Instrument] = {i.slug: i for i in ALL}
BY_SYMBOL: dict[str, Instrument] = {i.symbol: i for i in ALL}

GROUPS: dict[str, list[Instrument]] = {"sector": SECTORS, "index": INDICES}
